enzyme/Enzyme/MLIR/Implementations/LinalgAutoDiffOpInterfaceImpl.cpp

```diff
@@ -0,0 +1,237 @@
+    Region *adjointRegion = &adjoint.getBodyRegion();
+    int numInputsAdjoint = adjoint.getInputs().size();
+    Location loc = op->getLoc();
+    int numCaches = 0;
+    SmallVector<Value> pushCaches;
+
+    // std::function<std::pair<Value, Value>(Type)> hook = nullptr;
+    std::function<std::pair<Value, Value>(Type)> hook =
+        [newOpRegion, adjointRegion, loc, &numCaches = numCaches,
+         numInputsNewOp, numInputsAdjoint, &pushCaches = pushCaches](Type t) {
+          OpBuilder builder(newOpRegion);
+          Value pushCache = builder.create<enzyme::InitOp>(loc, t);
+          pushCaches.push_back(pushCache);
+          newOpRegion->addArgument(t, loc);
+
+          Value popCache = adjointRegion->insertArgument(
+              numInputsAdjoint + numCaches, t, loc);
+          numCaches++;
+          return std::pair<Value, Value>(pushCache, popCache);
+        };
+
+    gutils->Logic.differentiate(
+        gutils, *linalgOp.getBlock()->getParent(), adjoint.getBodyRegion(),
+        /*parentRegion=*/false, buildFuncReturnOp, hook);
+
+    // TODO add pushCaches to the yield in newOp
+    auto newOpYield = cast<linalg::YieldOp>(
+        cast<linalg::GenericOp>(newOp).getBodyRegion().front().getTerminator());
+    for (auto pc : pushCaches) {
+      newOpYield.getValuesMutable().append(pc);
+    }
+
+    Block *body = &(adjoint.getBodyRegion().front());
+    auto yieldOp = cast<linalg::YieldOp>(body->getTerminator());
+
+    for (auto opOperand : yieldOp.getOperands()) {
+      body->addArgument(opOperand.getType(), opOperand.getLoc());
+    }
+    OpBuilder builderAdd(yieldOp);
+    for (auto it : llvm::enumerate(yieldOp.getOperands())) {
+      Value arg = body->getArgument(outputs.size() + numCaches + it.index());
+      auto diffeType = cast<AutoDiffTypeInterface>(arg.getType());
+      Value grad = diffeType.createAddOp(builderAdd, it.value().getLoc(), arg,
+                                         it.value());
+      yieldOp.setOperand(it.index(), grad);
+    }
+
+    auto indexing_maps = newOp.getIndexingMapsArray();
+    auto indexing_maps_adjoint = adjoint.getIndexingMapsArray();
+    for (int i = 0; i < numCaches; i++) {
+      Value cacheArg = body->getArgument(outputs.size() + i);
+
+      Type ct = cacheArg.getType();
+      Type type = MemRefType::get(ArrayRef(shapes), ct);
+      Value alloc = cacheBuilder.create<memref::AllocOp>(
+          op->getLoc(), type, ValueRange(iterationDomains));
+      Value cache = gutils->initAndPushCache(alloc, cacheBuilder);
+      alloc.getDefiningOp()->setAttr("operand_segment_sizes",
+                                     cacheBuilder.getDenseI32ArrayAttr({1, 0}));
+
+      cast<linalg::GenericOp>(newOp).getOutputsMutable().append(
+          ValueRange({alloc}));
+      indexing_maps.push_back(AffineMap::getMultiDimIdentityMap(
+          iterationDomains.size(), cacheBuilder.getContext()));
+
+      OpBuilder builder2(adjoint);
```

Comment:
It is usually a bad practice to have separate builders. Prefer one builder and use `OpBuilder::InsertionGuard` RAII object to control insertion points.